internal/xds/rbac/rbac_engine_test.go

```diff
@@ -1007,85 +1221,704 @@ func (s) TestChainEngine(t *testing.T) {
+							return
+						}
+						connCh <- conn
+					}()
+					_, err = net.Dial("tcp", lis.Addr().String())
+					if err != nil {
+						t.Fatalf("Error dialing: %v", err)
+					}
+					conn := <-connCh
+					defer conn.Close()
+					getConnection = func(context.Context) net.Conn {
+						return conn
+					}
+					ctx = peer.NewContext(ctx, data.rpcData.peerInfo)
+					stream := &ServerTransportStreamWithMethod{
+						method: data.rpcData.fullMethod,
+					}
+
+					ctx = grpc.NewContextWithServerTransportStream(ctx, stream)
+					err = cre.IsAuthorized(ctx)
+					if gotCode := status.Code(err); gotCode != data.wantStatusCode {
+						t.Fatalf("IsAuthorized(%+v, %+v) returned (%+v), want(%+v)", ctx, data.rpcData.fullMethod, gotCode, data.wantStatusCode)
+					}
+					if !reflect.DeepEqual(b.auditEvents, data.wantAuditEvents) {
+						t.Fatalf("Unexpected audit event for query:%v", data)
+					}
+
+					// this builder's auditEvents can be shared for several queries, make sure it's empty
+					b.auditEvents = nil
+				}()
+			}
+		})
+	}
+}
+
+type ServerTransportStreamWithMethod struct {
+	method string
+}
+
+func (sts *ServerTransportStreamWithMethod) Method() string {
+	return sts.method
+}
+
+func (sts *ServerTransportStreamWithMethod) SetHeader(md metadata.MD) error {
+	return nil
+}
+
+func (sts *ServerTransportStreamWithMethod) SendHeader(md metadata.MD) error {
+	return nil
+}
+
+func (sts *ServerTransportStreamWithMethod) SetTrailer(md metadata.MD) error {
+	return nil
+}
+
+// An audit logger that will log to the auditEvents slice
+type TestAuditLoggerBuffer struct {
+	auditEvents *[]*audit.Event
+}
+
+func (logger *TestAuditLoggerBuffer) Log(e *audit.Event) {
+	*(logger.auditEvents) = append(*(logger.auditEvents), e)
+}
+
+// Builds TestAuditLoggerBuffer
+type TestAuditLoggerBufferBuilder struct {
+	auditEvents []*audit.Event
+	testName    string
+}
+
+// The required config for TestAuditLoggerBuffer
+type TestAuditLoggerBufferConfig struct {
+	audit.LoggerConfig
+}
+
+func (b *TestAuditLoggerBufferBuilder) ParseLoggerConfig(configJSON json.RawMessage) (config audit.LoggerConfig, err error) {
+	return TestAuditLoggerBufferConfig{}, nil
+}
+
+func (b *TestAuditLoggerBufferBuilder) Build(config audit.LoggerConfig) audit.Logger {
+	return &TestAuditLoggerBuffer{auditEvents: &b.auditEvents}
+}
+
+func (b *TestAuditLoggerBufferBuilder) Name() string {
+	return b.testName + "_TestAuditLoggerBuffer"
+}
+
+// An audit logger to test using a custom config
+type TestAuditLoggerCustomConfig struct{}
+
+func (logger *TestAuditLoggerCustomConfig) Log(*audit.Event) {}
+
+// Build TestAuditLoggerCustomConfig
+type TestAuditLoggerCustomConfigBuilder struct {
+	testName string
+}
+
+// The custom config for the TestAuditLoggerCustomConfig logger
+type TestAuditLoggerCustomConfigConfig struct {
+	audit.LoggerConfig
+	Abc int
+	Xyz string
+}
+
+// Parses TestAuditLoggerCustomConfigConfig
+// Hard-coded to match with it's test case above
```

Comment:
```suggestion
// Parses TestAuditLoggerCustomConfigConfig. Hard-coded to match with it's test
// case above.
```
line wrap + punctuation pls